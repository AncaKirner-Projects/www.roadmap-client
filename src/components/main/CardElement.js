import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import AlertDialog from './AlertDialog';
import ActionTypes from '../../store/actions/actionTypes';
import { addToCart, deleteCartStatusMessage } from '../../store/actions/cartActions';

const CardStyle = styled(Card)`
  width: 310px;
  height: 320px;
  display: inline-block;
  margin: 25px 0px 0px 25px;
`;

const TextStyle = styled(Typography)`
  height: 110px
`;

const CardActionsStyle = styled(CardActions)`
  justify-content: center;
`;

class CardElement extends Component {
  state = {
    openAlertDialog: false
  }

  handleAlertDialogClose = () => {
    this.props.deleteCartStatusMessage(ActionTypes.ADD_TO_CART);
    this.setState({ openAlertDialog: false });
  }

  handleAddToCart = () => {
    this.props.addToCart(this.props.product);
    this.setState({ openAlertDialog: true });
  }

  render() {
    const {
      id, prod_name: name, prod_description: description, price
    } = this.props.product;
    const { status } = this.props;
    const renderLink = itemProps => <Link to={`/products/${id}`} {...itemProps} />;
    const addProductStatus = status && status.find(elem => elem.type === ActionTypes.ADD_TO_CART);

    return (
      <CardStyle>
        <CardContent>
          <TextStyle align="center" variant="h5" to={`/products/${id}`} component={renderLink}>
            {name}
          </TextStyle>
          <TextStyle align="center" component="p">
            {description}
          </TextStyle>
          <Typography align="center" variant="h5" component="h1" color="primary">
            {price}
            RON
          </Typography>
          <CardActionsStyle>
            <Button variant="contained" size="small" color="primary" onClick={this.handleAddToCart}>Add to cart</Button>
          </CardActionsStyle>
          <AlertDialog
            open={this.state.openAlertDialog}
            handleClose={this.handleAlertDialogClose}
            title="Request Status"
            text={addProductStatus && addProductStatus.message}
          />
        </CardContent>
      </CardStyle>
    );
  }
}

const mapStateToProps = state => ({
  cart: state.cart ? state.cart.products : [],
  status: state.cart ? state.cart.status : []
});

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators({
    addToCart,
    deleteCartStatusMessage
  }, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(CardElement);
