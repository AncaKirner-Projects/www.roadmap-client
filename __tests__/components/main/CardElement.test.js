import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import Button from '@material-ui/core/Button';
import AlertDialog from '../../../src/components/main/AlertDialog';
import CardElement from '../../../src/components/main/CardElement';
import ActionTypes from '../../../src/store/actions/actionTypes';

const UnwrappedCard = CardElement.WrappedComponent;
const mockStore = configureMockStore();

describe('<CardElement />', () => {
  const store = mockStore({
    cart: {
      status: [
        {
          type: ActionTypes.ADD_TO_CART,
          message: 'Test message'
        }
      ]
    }
  });

  const product = {
    prod_name: 'Test Product',
    prod_description: 'Product description',
    price: 23.6
  };
  const props = {
    addToCart: jest.fn(),
    deleteCartStatusMessage: jest.fn()
  };

  it('renders an element card', () => {
    const wrapper = shallow(<CardElement product={product} store={store} />);
    const component = wrapper.dive();

    expect(toJson(component)).toMatchSnapshot();
  });

  it('handleAddToCart', () => {
    const wrapper = shallow(<UnwrappedCard {...props} product={product} />);
    wrapper.find(Button).simulate('click');

    expect(props.addToCart).toHaveBeenCalled();
  });

  it('handleAlertDialogClose', () => {
    const wrapper = shallow(<UnwrappedCard {...props} product={product} />);
    const component = wrapper.instance();
    component.handleAlertDialogClose = jest.fn();

    wrapper.find(AlertDialog).simulate('handleClose');

    expect(props.deleteCartStatusMessage).toHaveBeenCalled();
  });
});
