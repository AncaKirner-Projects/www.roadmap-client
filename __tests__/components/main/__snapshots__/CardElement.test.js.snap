// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardElement /> renders an element card 1`] = `
<Styled(WithStyles(Card))>
  <WithStyles(CardContent)>
    <Styled(WithStyles(Typography))
      align="center"
      component={[Function]}
      to="/products/undefined"
      variant="h5"
    >
      Test Product
    </Styled(WithStyles(Typography))>
    <Styled(WithStyles(Typography))
      align="center"
      component="p"
    >
      Product description
    </Styled(WithStyles(Typography))>
    <WithStyles(Typography)
      align="center"
      color="primary"
      component="h1"
      variant="h5"
    >
      23.6
      RON
    </WithStyles(Typography)>
    <Styled(WithStyles(CardActions))>
      <WithStyles(Button)
        color="primary"
        onClick={[Function]}
        size="small"
        variant="contained"
      >
        Add to cart
      </WithStyles(Button)>
    </Styled(WithStyles(CardActions))>
    <AlertDialog
      onHandleClose={[Function]}
      open={false}
      text="Test message"
      title="Request Status"
    />
  </WithStyles(CardContent)>
</Styled(WithStyles(Card))>
`;
